[Relay, TOPI] Add negative log likelihood loss (nll_loss) op

zhuzilin opened the pull request:
This PR adds the `nll_loss` op to relay and topi, so that we could translate `aten::nll_loss` in pytorch frontend. The `nll_loss` is the underlying function for cross entropy in pytorch and is very important in the way of supporting training in tvm.

Thank you for your time on reviewing this PR :).

tkonolige commented on include/tvm/topi/nn.h:
```diff
@@ -642,6 +643,54 @@ inline tvm::te::Tensor batch_to_space_nd(const tvm::te::Tensor& data,
+ * \return A Tensor whose op member is the batch_to_space_nd operation
+ */
+inline Tensor nll_loss(const Tensor& input, const Tensor& target, const Tensor& weight,
+                       std::string reduction = "mean", int ignore_index = -100,
```
What is the significance of the -100 for `ignore_index`?

altanh replied:
I believe this is the default used in PyTorch, there's no inherent significance to it (although certain data loading/preprocessing scripts seem to try and match this value - e.g. I've see this in certain BERT training data loaders)

zhuzilin replied:
Yes, it's the default value used in pytorch, which is mainly a random negative number in my opinion.

tkonolige commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -2973,6 +2973,34 @@ def cross_entropy_with_logits(predictions, targets):
 
 
+def nll_loss(input, target, weight, reduction="mean", ignore_index=-100):
+    """Negative log likelihood loss.
```
Can you document how `nll_loss` is computed (i.e. a mathematical expression).

tkonolige commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -2973,6 +2973,34 @@ def cross_entropy_with_logits(predictions, targets):
+      The weight of each target value.
+
+    reduction : string
+      The reduction method to apply to the output.
```
Can you list the possible options for reduction.

tkonolige commented on python/tvm/topi/nn/loss.py:
```diff
@@ -0,0 +1,53 @@
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name,unused-argument
+"""TVM operator negative log likelihood loss compute."""
```
We may add other losses in the future, so maybe this should read something like "Loss functions definitions".

tkonolige commented on src/relay/op/nn/nn.cc:
```diff
@@ -1091,6 +1092,60 @@ Accept logits.
+
+bool NLLLossRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
+                const TypeReporter& reporter) {
+  ICHECK_EQ(types.size(), 4);
```
```suggestion
  ICHECK_EQ(types.size(), 4) << "NLLLossRel expects 4 types, but " << types.size() << " were provided.";
```

tkonolige commented on src/relay/op/nn/nn.cc:
```diff
@@ -1091,6 +1092,60 @@ Accept logits.
+      << "NLLLossRel: input should be one dimension larger than target, "
+      << "input shape = " << input->shape << ", "
+      << "target shape = " << target->shape;
+  ICHECK(weight->shape.size() == 1);
```
Please add an error message.

tkonolige commented on src/relay/op/nn/nn.cc:
```diff
@@ -1091,6 +1092,60 @@ Accept logits.
+      << "which is the length of weight, "
+      << "input shape = " << input->shape << ", "
+      << "weight shape = " << weight->shape;
+  ICHECK(input->dtype == weight->dtype && input->dtype.is_float());
+  ICHECK(target->dtype.is_int());
```
Please add error messages

tkonolige reviewed (changes requested):
Thanks for submitting this op. I know some people are looking forward to having it available. I've left some comments for things to fix (mostly around documentation).

zhuzilin commented:
@tkonolige 
Thank you for your review! I've added the docs. And because the lint check doesn't allow parameter name as `input`, I renamed the parameters to be the same form as `cross_entropy` and `cross_entropy_with_logits` (the old one was the same as `aten::nll_loss`).

tkonolige commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -2973,6 +2973,40 @@ def cross_entropy_with_logits(predictions, targets):
+def nll_loss(predictions, targets, weights, reduction="mean", ignore_index=-100):
+    """Negative log likelihood loss.
+
+    output{n, i_1, i_2, ..., i_k} = predictions{n, t, i_1, i_2, i_k}
```
How does weights and ignore_index factor in here?

altanh replied:
missing a minus sign in the math here?

altanh commented on include/tvm/topi/nn.h:
```diff
@@ -642,6 +643,54 @@ inline tvm::te::Tensor batch_to_space_nd(const tvm::te::Tensor& data,
+ * \param name The name of the operation.
+ * \param tag The tag to mark the operation.
+ *
+ * \return A Tensor whose op member is the batch_to_space_nd operation
```
fix incorrect return docstring

altanh commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -2973,6 +2973,40 @@ def cross_entropy_with_logits(predictions, targets):
+    targets : tvm.relay.Expr
+      The target value of each prediction.
+
+    weights : tvm.relay.Expr
```
can we make weights optional, like PyTorch? weights=1 is a pretty common case I believe and we could add a fast path implementation that skips the scaling

zhuzilin replied:
@altanh We can make weights an optional parameter. I wonder if there are any example of a relay op with an optional tensor parameter that I can learn from. And also, how should we deal with gradient of an optional parameter? BTW, is there any better way we can mark a parameter as "no need for gradient" instead of returning an `one_like` grad?

altanh replied:
Hmm not sure, that's a good point- let's just keep the weights for now. As for not needing a gradient, currently there is no other way than just putting some dummy value. It might make sense for us to introduce a `stop_gradient` dummy op which cuts the gradient computation from going further at undifferentiable arguments. Thanks!

altanh commented on src/relay/op/nn/nn.cc:
```diff
@@ -1091,6 +1092,65 @@ Accept logits.
+      << "weights shape = " << weights->shape;
+  ICHECK(predictions->dtype == weights->dtype && predictions->dtype.is_float())
+      << "NLLLossRel: predictions and weights should be of the same floating type.";
+  ICHECK(targets->dtype.is_int()) << "NLLLossRel: targets should be of int type.";
```
can you replace these ICHECKs with Diagnostics?

zhuzilin replied:
Is there a plan on changing all `ICHECKS` to diagnostics? Or any rules on when to use `ICHECK` and when to use diagnostics?

altanh replied:
basically, if the error can happen due to user input (e.g. using wrong shapes), we should definitely use diagnostics. ICHECK should be reserved only for internal compiler checks that should basically never fail unless there's a bug somewhere. The diagnostic framework is fairly new so a lot of old code still uses ICHECK incorrectly, we just need to slowly go through and update them unfortunately

altanh commented on python/tvm/topi/nn/loss.py:
```diff
@@ -0,0 +1,58 @@
+def nll_loss(predictions, targets, weights, reduction, ignore_index):
+    """Negative log likelihood loss on the input data.
+
+    output{n, i_1, i_2, ..., i_k} = predictions{n, t, i_1, i_2, i_k}
```
missing minus sign in math? and weights

altanh commented on include/tvm/topi/nn.h:
```diff
@@ -642,6 +643,54 @@ inline tvm::te::Tensor batch_to_space_nd(const tvm::te::Tensor& data,
+      [&](const tvm::Array<tvm::tir::Var>& target_indices) {
+        auto c = targets(target_indices);
+        tvm::Array<tvm::PrimExpr> pred_indices;
+        for (size_t i = 0; i < target_indices.size(); i++) {
+          pred_indices.push_back(target_indices[i]);
+          if (i == 0) {
+            pred_indices.push_back(c);
+          }
+        }
```
This is clearer to me
```suggestion
        pred_indices.push_back(target_indices[0]);  // batch index
        pred_indices.push_back(c); // class index
        for (size_t i = 1; i < target_indices.size(); i++) {
          pred_indices.push_back(target_indices[i]);  // indices for multidimensional loss
        }
```

altanh commented on include/tvm/topi/nn.h:
```diff
@@ -642,6 +643,54 @@ inline tvm::te::Tensor batch_to_space_nd(const tvm::te::Tensor& data,
+ */
+inline Tensor nll_loss(const Tensor& predictions, const Tensor& targets, const Tensor& weights,
+                       std::string reduction = "mean", int ignore_index = -100,
+                       const std::string name = "nll_loss", const std::string tag = kBroadcast) {
```
should the tag be `kOpaque` to match the Relay pattern?

zhuzilin replied:
@altanh I am confused with the `tag` in topi (ones in `topi/tags.h`) and the `OpPatternKind` in `relay/op_attr_types.h`. It seems that they are not matched. Could you tell me the usage of them in tvm? Thank you~

altanh replied:
I see, that is confusing.. I'll get back to you on this soon

altanh replied:
> @altanh Could you take a look at the tests? Thank you~ And I wonder if there is any update with the `tag` and `OpPatternKind`?

I don't have much of an update for the tag, maybe you could try leaving it empty string?

altanh reviewed (changes requested):
Thanks for the PR, excited to see others working on training support! I left some requests. A few more things:
- please add unit tests for the op
- are you planning on registering a gradient for this operator?

zhuzilin commented:
@tkonolige Sorry about the incomplete formula... I don't know what  I was thinking about yesterday :scream:

altanh commented:
> Could you tell me how to decide to level of an op? And I could add the test to the correct place.

Yep, sorry that the documentation of support level is lacking! We do have a doc at (https://tvm.apache.org/docs/langref/relay_op.html#overview-of-operators) but it seems outdated and missing quite a bit, so this needs to be improved. For this op, I think we can go with level 10 which matches the existing `nn.cross_entropy` op. I think we might want to remove this operator in the future since NLLLoss seems to be more general and subsumes it in my opinion, but that will be a separate PR :)

Thanks!

zhuzilin commented:
@altanh Thanks. I'll add the test soon. Could you also check the comments above relates to the optional weight, Diagnostics and the tag?

altanh commented:
> @altanh Thanks. I'll add the test soon. Could you also check the comments above relate to the optional weight, Diagnostics and the tag?

replied

zhuzilin commented:
@altanh @tkonolige Sorry for the late update. I've just added the test of `nll_loss` for relay, topi and pytorch frontend. Could you have another look?

zhuzilin commented:
@altanh Could you take a look at the tests? Thank you~ And I wonder if there is any update with the `tag` and `OpPatternKind`?

altanh commented:
It looks like we have some CI problems currently

altanh commented on tests/python/topi/python/test_topi_loss.py:
```diff
@@ -0,0 +1,70 @@
+    def check_device(target, dev):
+        print("Running on target: %s" % target)
+        with tvm.target.Target(target):
+            s = tvm.topi.testing.get_injective_schedule(target)(nll_loss_result)
```
I'm not too familiar with the scheduling workflow, but is it right for this to be an injective schedule? cc @tkonolige

tkonolige replied:
Try
```suggestion
            s = tvm.te.create_schedule(nll_loss_result.op)
```

zhuzilin commented:
@altanh @tkonolige @tqchen Is there anything I can help with this pr... It has been stuck for a while....

tkonolige commented on python/tvm/relay/op/nn/_nn.py:
```diff
@@ -886,6 +886,17 @@ def compute_cross_entropy_with_logits(attrs, inputs, out_dtype):
+
+
+reg.register_reduce_schedule("nn.nll_loss")
+reg.register_pattern("nn.nll_loss", OpPattern.OPAQUE)
```
```suggestion
reg.register_pattern("nn.nll_loss", OpPattern.OUT_ELEMWISE_FUSABLE)
```

I believe you can fuse into the out here.

tkonolige commented on tests/python/topi/python/test_topi_loss.py:
```diff
@@ -0,0 +1,70 @@
+
+
+@tvm.testing.uses_gpu
+def test_nll_loss():
```
Can you refactor this function to use tvm.testing.parametrize_targets (https://github.com/apache/tvm/blob/main/python/tvm/testing.py#L704)?

zhuzilin replied:
Refactored.

tkonolige reviewed (changes requested):
Sorry for being slow on this. Could you make just a couple more changes.

zhuzilin commented:
@tkonolige Thank you for the reviews! I've modified the parts based on them. Could you take another look?

tkonolige reviewed (changes requested):
One last change and then I think this will be good to go.

zhuzilin commented:
> One last change and then I think this will be good to go.

@tkonolige Could you point out the change that I need to make? Thank you~

tkonolige commented on tests/python/relay/test_op_level10.py:
```diff
@@ -577,6 +577,49 @@ def _verify(input_shape, diagonal_shape, dtype, k=0, align="RIGHT_LEFT"):
 
 
+@tvm.testing.uses_gpu
+def test_nll_loss():
```
Switch this to use `parameterize_targets`.

tkonolige reviewed (changes requested):
Sorry, I forgot to include the actual change I wanted.

zhuzilin commented:
@tkonolige I've updated the test. But there seems to be some error with the cuda target. Could you give me some help? Thank you! Part of the error message in the CI is listed below:

>dev = cuda(0), target = 'cuda'
>
>
>```python
>    @tvm.testing.parametrize_targets
>
>    def test_nll_loss(dev, target):
>
>>       verify_nll_loss(dev, target, (10, 5))
>```
>
>
>tests/python/topi/python/test_topi_loss.py:60: 
>
>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>
>tests/python/topi/python/test_topi_loss.py:40: in verify_nll_loss
>
>    fn = tvm.build(s, [predictions, targets, weights, nll_loss_result], target, name="nll_loss")
>
>python/tvm/driver/build_module.py:353: in build
>
>    mod_host, mdev = _build_for_device(input_mod, tar, target_host)
>
>python/tvm/driver/build_module.py:177: in _build_for_device
>
>    mod_mixed = tvm.transform.Sequential(opt_mixed)(mod_mixed)
>
>python/tvm/ir/transform.py:161: in __call__
>
>    return _ffi_transform_api.RunPass(self, mod)
>...
>
>E     Did you forget to bind?
>
>E       Variable `T_divide` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `targets` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `weights` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `targets` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `targets` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `weights` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `targets` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `predictions` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E       Variable `targets` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
>
>E     File "/workspace/src/tir/analysis/verify_memory.cc", line 202
>
>E   RuntimeError: Memory verification failed with the following errors:
>
>E   PrimFunc([predictions, targets, weights, T_divide]) attrs={"global_symbol": "nll_loss", "tir.noalias": (bool)1, "target": cuda -keys=cuda,gpu -max_num_threads=1024 -thread_warp_size=32} {
>
>E     // attr [nll_loss] storage_scope = "global"
>
>E     allocate nll_loss[float32 * 10]
>
>E     // attr [nll_loss_red] storage_scope = "global"
>
>E     allocate nll_loss_red[float32 * 1]
>
>E     // attr [nll_loss_red] storage_scope = "global"
>
>E     allocate nll_loss_red[float32 * 1]
>
>E     for (ax0, 0, 10) {
>
>E       nll_loss[ax0] = tir.if_then_else((targets[ax0] != -100), ((0f - predictions[((ax0*5) + targets[ax0])])*weights[targets[ax0]]), 0f)
>
>E     }
>
>E     nll_loss_red[0] = 0f
>
>E     for (k0, 0, 10) {
>
>E       nll_loss_red[0] = (nll_loss_red[0] + nll_loss[k0])
>
>E     }
>
>E     for (ax0, 0, 10) {
>
>E       nll_loss[ax0] = tir.if_then_else((targets[ax0] != -100), weights[targets[ax0]], 0f)
>
>E     }
>
>E     nll_loss_red[0] = 0f
>
>E     for (k0, 0, 10) {
>
>E       nll_loss_red[0] = (nll_loss_red[0] + nll_loss[k0])
>
>E     }
>
>E     T_divide[0] = (nll_loss_red[0]/nll_loss_red[0])
>
>E   }

zhuzilin commented:
@tkonolige Could you take another look? Thank you~

tqchen commented:
cc @altanh @vinx13 it would be great if you can help take another look

vinx13 reviewed (commented):
topi test on coda didn't pass. Looks like there are some issues with scheduling, when you called `register_reduction_schedule`,on cuda target this schedule is used https://github.com/apache/tvm/blob/main/python/tvm/topi/cuda/reduction.py#L26 could you check if there are anything wrong?

altanh commented:
> @tkonolige I've updated the test. But there seems to be some error with the cuda target. Could you give me some help? Thank you! Part of the error message in the CI is listed below:

I think this error is due to using the default schedule (from `te.create_schedule`) which does not bind the iteration axes to the GPU threads. This is a bit outside of area of understanding but the solution will be to use the respective cuda/gpu schedules for each compute in the kernel (e.g. `schedule_reduce`, `schedule_injective`, etc.). Maybe @vinx13 can give some more detailed help?

vinx13 commented on tests/python/topi/python/test_topi_loss.py:
```diff
@@ -0,0 +1,69 @@
+    nll_loss_result = topi.nn.nll_loss(predictions, targets, weights, reduction, ignore_index)
+
+    with tvm.target.Target(target):
+        s = tvm.te.create_schedule(nll_loss_result.op)
```
default schedule is used here which caused ci errors
```suggestion
        fschedule = tvm.topi.testing.get_reduce_schedule(target)
        s = fschedule([nll_loss_result])
```

zhuzilin commented:
@vinx13 @altanh Thank you for your help!

> tag here is topi-level, sometimes we use it to identify a specific compute operation during schedule, otherwise we can leave it empty

If I change the value of `tag` to an empty string, it will fail the check in `schedule_reduce`, which is:
https://github.com/apache/tvm/blob/d0791d3db971a111826d96201bd1e4c9c0d531da/python/tvm/topi/x86/reduction.py#L84

I'm not sure if I need to adjust somewhere else...

vinx13 commented:
> @vinx13 @altanh Thank you for your help!
> 
> > tag here is topi-level, sometimes we use it to identify a specific compute operation during schedule, otherwise we can leave it empty
> 
> If I change the value of `tag` to an empty string, it will fail the check in `schedule_reduce`, which is:
> 
> https://github.com/apache/tvm/blob/d0791d3db971a111826d96201bd1e4c9c0d531da/python/tvm/topi/x86/reduction.py#L84
> 
> I'm not sure if I need to adjust somewhere else...

in this case we need to change tag to `kInjective` as the reduction op is not broadcast

zhuzilin commented:
> in this case we need to change tag to `kInjective` as the reduction op is not broadcast

@vinx13 Changing tag to `kInjective` will fail the tag check in `traverse_after_reduce` in
https://github.com/apache/tvm/blob/d0791d3db971a111826d96201bd1e4c9c0d531da/python/tvm/topi/x86/reduction.py#L94
whereas the empty tag was triggering `traverse_before_reduce`....

It's worth noticing that the tag is of:

```c++
  auto T = tvm::te::compute(
      targets->shape,
      ...
      name, tag);
```

which is an element-wise operation on `targets`. And when `reduction="mean"` or `reduction="sum"`, `T` will be reduced, whereas when `reduction="none"`, the `T` will be returned as the result of the `nll_loss`.

Therefore, as `nll_loss` is registered to use reduce schedule, when `reduction="mean"` or `"sum"`, the `T` will go through the checks in `traverse_before_reduce` while when `reduction="none"`, `T` will go through checks in `traverse_after_reduce`. Right now, the `kBroadcast` happens to be the only tag to satisfiy both checks and passed the CI . To really solve this problem, I'm afraid we may need to define different tag for different op attr...

vinx13 reviewed (approved):
LGTM

vinx13 commented:
This is merged, thanks @zhuzilin @altanh @tkonolige @tqchen

zhuzilin commented:
@vinx13 @altanh @tkonolige @tqchen Thank you for your help! I'll submit a pr for the gradient of nll_loss soon.
